### PR TITLE
Debug info selection

### LINKED ISSUE #11858: Print code snippets next to source locations in IR
With `--asm` option the compiler prints not only source locations but also small snippets of the Solidity code from those locations. For example:

```
tag_5:
    /* "test.sol":111:117  uint x */
  0x00
    /* "test.sol":120:121  1 */
  0x01
    /* "test.sol":111:121  uint x = 1 */
  swap1
  pop
    /* "test.sol":101:128  {... */
  pop
    /* "test.sol":81:128  function f() public {... */
  jump      // out
```

In IR we only get locations:
```yul
/// @src 0:81:128
function fun_f_21() {

    /// @src 0:120:121
    let expr_18 := 0x01
    /// @src 0:111:121
    let var_x_17 := convert_t_rational_1_by_1_to_t_uint256(expr_18)

}
/// @src 0:0:130

function identity(value) -> ret {
    ret := value
}
```

`AsmPrinter` should be extended to add snippets in IR too.

### PR DESCRIPTION
**Pull request by cameel**
Fixes #11858.
~Depends on #11867.~ Merged.
~Weakly depends on #11993 - `chk_codingstyle` fails because of trailing whitespace without it.~ Merged
~Weakly depends on #11991 - I have one empty input file in cmdline tests, which makes `xxx_cli` jobs fail.~
Now it actually depends on #11991.

I could work around the dependency on #11991 and #11993 but they'll likely go in first anyway.

Still work in progress:
- [x] Make it affect `--asm` output as well.
- [x] Error when used in input modes that do not need it.
- [x] Standard JSON setting.
- [x] Command-line tests.
- [x] Documentation + changelog.

### CODE REVIEW

**Comment by chriseth** [comment]
I think we don't need to change the --asm output here.

**Comment by cameel** [comment]
Oh, well, I just finished making it support that option. I thought that the option was supposed to affect the EVM assembly too?

**Comment by chriseth** [comment]
If it's not a big deal, we can also add it there.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `52a625e9383778af08901b9de3ce5fef2d4fdaf1`:
```
Error: Trailing whitespace found:
test/cmdlineTests/debug_info_in_yul_and_evm_asm_print_location_only/output:183:            /// @src 0:60:101  
test/cmdlineTests/debug_info_in_yul_and_evm_asm_print_location_only/output:195:                /// @src 0:60:101  
test/cmdlineTests/standard_debug_info_in_yul_and_evm_asm_print_location_only/output.json:188:            /// @src 0:60:101  
test/cmdlineTests/standard_debug_info_in_yul_and_evm_asm_print_location_only/output.json:200:                /// @src 0:60:101  
test/cmdlineTests/strict_asm_debug_info_print_all/output:8:        /// @src 0:60:101  
test/cmdlineTests/strict_asm_debug_info_print_all/output:11:        /// @src 0:77:99  
test/cmdlineTests/strict_asm_debug_info_print_location_only/output:8:        /// @src 0:60:101  
test/cmdlineTests/strict_asm_debug_info_print_location_only/output:11:        /// @src 0:77:99  

```
Please check that your changes are working as intended.

**Comment by cameel** [comment]
This is now ready for review.

File: liblangutil/DebugInfoSelection.h
Comment (cameel) [inline liblangutil/DebugInfoSelection.h:40]: This might look a bit over-engineered on its own but I'm planning to make it a template and reuse it for other, similar options. It's based on `SolverChoice`, which already has most of this functionality. We have a lot of CLI options that represent a bunch of boolean flags parsed from comma-separated lists:
- [`SolverChoice`](https://github.com/ethereum/solidity/blob/v0.8.7/libsmtutil/SolverInterface.h#L39-L107)
- [`ModelCheckerEngine`](https://github.com/ethereum/solidity/blob/v0.8.7/libsolidity/formal/ModelCheckerSettings.h#L115-L142)
- [`ModelCheckerTargets`](https://github.com/ethereum/solidity/blob/v0.8.7/libsolidity/formal/ModelCheckerSettings.h#L92-L113)
- [`CombinedJsonRequests`](https://github.com/ethereum/solidity/blob/v0.8.7/solc/CommandLineParser.h#L76-L98)
- [`CompilerOutputs`](https://github.com/ethereum/solidity/blob/v0.8.7/solc/CommandLineParser.h#L54-L74)

Mine is the sixth and we might want to add even more in the future.

A big problem with all of them is that when adding a new flag, forgetting to include it in one place will not cause an error, just give you wrong results. My version solves it by defining a mapping between names and members and reusing it in all definitions that need to handle all members.

File: liblangutil/DebugInfoSelection.h
Comment (cameel) [inline liblangutil/DebugInfoSelection.h:45]: Should `@ast-id` be enabled by default?

File: libsolidity/interface/StandardCompiler.cpp
Comment (cameel) [inline libsolidity/interface/StandardCompiler.cpp:838]: I could print snippet without location just fine but I don't think we want to have to handle this case in the parser.

File: libyul/AssemblyStack.h
Comment (cameel) [inline libyul/AssemblyStack.h:85]: I could avoid having to change all the places that create `AssemblyStack` by giving this a default but it also makes it very easy to forget to pass it in, which could bypass the CLI/JSON setting.

File: solc/CommandLineParser.cpp
Comment (cameel) [inline solc/CommandLineParser.cpp:70]: In the end I called it `--debug-info` rather than `--ir-debug-info` because it affects EVM assembly as well.

File: test/cmdlineTests/debug_info_in_yul_and_evm_asm_print_none/output
Comment (cameel) [inline test/cmdlineTests/debug_info_in_yul_and_evm_asm_print_none/output:62]: `--debug-info none` does not suppress `@use-src`. Should it?

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `11b89e45807c9643eb992e6af4a4611cbf0f9bf1`:
```
Error: Trailing whitespace found:
test/cmdlineTests/debug_info_in_yul_and_evm_asm_print_location_only/output:183:            /// @src 0:60:101  
test/cmdlineTests/debug_info_in_yul_and_evm_asm_print_location_only/output:195:                /// @src 0:60:101  
test/cmdlineTests/standard_debug_info_in_yul_and_evm_asm_print_location_only/output.json:188:            /// @src 0:60:101  
test/cmdlineTests/standard_debug_info_in_yul_and_evm_asm_print_location_only/output.json:200:                /// @src 0:60:101  
test/cmdlineTests/strict_asm_debug_info_print_all/output:8:        /// @src 0:60:101  
test/cmdlineTests/strict_asm_debug_info_print_all/output:11:        /// @src 0:77:99  
test/cmdlineTests/strict_asm_debug_info_print_location_only/output:8:        /// @src 0:60:101  
test/cmdlineTests/strict_asm_debug_info_print_location_only/output:11:        /// @src 0:77:99  

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `9e487e26ea6af164ca57a254f5a786944ff0ffbe`:
```
Error: Trailing whitespace found:
test/cmdlineTests/debug_info_in_yul_and_evm_asm_print_location_only/output:183:            /// @src 0:60:101  
test/cmdlineTests/debug_info_in_yul_and_evm_asm_print_location_only/output:195:                /// @src 0:60:101  
test/cmdlineTests/standard_debug_info_in_yul_and_evm_asm_print_location_only/output.json:188:            /// @src 0:60:101  
test/cmdlineTests/standard_debug_info_in_yul_and_evm_asm_print_location_only/output.json:200:                /// @src 0:60:101  
test/cmdlineTests/strict_asm_debug_info_print_all/output:8:        /// @src 0:60:101  
test/cmdlineTests/strict_asm_debug_info_print_all/output:11:        /// @src 0:77:99  
test/cmdlineTests/strict_asm_debug_info_print_location_only/output:8:        /// @src 0:60:101  
test/cmdlineTests/strict_asm_debug_info_print_location_only/output:11:        /// @src 0:77:99  

```
Please check that your changes are working as intended.

**Comment by leonardoalt** [comment]
Needs rebasing

File: liblangutil/DebugInfoSelection.cpp
Comment (Marenz) [inline liblangutil/DebugInfoSelection.cpp:20]: extra blankline

**Review by Marenz — CHANGES_REQUESTED** [review]
LGTM. Just a tiny whitespace change.

File: liblangutil/DebugInfoSelection.cpp
Comment (cameel) [reply liblangutil/DebugInfoSelection.cpp:20]: This one is on purpose. It's the header corresponding to the .cpp file so it should be the first include and is technically not in the same "group". I.e. I wouldn't sort it alphabetically with other includes from the same dir.

**Comment by cameel** [comment]
Note: this is now based on the CI fix (#12106). Please don't merge until that fix gets merged.

**Review by Marenz — APPROVED** [review]

File: test/cmdlineTests/debug_info_in_yul_and_evm_asm_print_none/output
Comment (chriseth) [reply test/cmdlineTests/debug_info_in_yul_and_evm_asm_print_none/output:62]: If we suppress it, we get yul source locations instead of empty source locations, so I think it's better this way.

**Pull request merged**
